### ISSUE TITLE
[DACON 반도체 박막 두께 분석 경진대회 1등]

### ISSUE THREAD

**Issue by 98hyun** [Sejong-Kaggle-Challengers/MAIN#11]
[링크](https://dacon.io/competitions/official/235554/codeshare/651?page=1&dtype=recent&ptype=pub)

**Comment by mingxoxo** [comment]
<h2>itertools</h2>
python에서 제공하는 자신만의 반복자를 만드는 모듈
APL, Haskell 및 SML의 구성 요소에서 영감을 받은 반복기 빌딩 블록을 구현하며 각각은 파이썬에 적합한 형태로 재 작성

`import itertools`
- [반복자(iterator)](https://ko.wikipedia.org/wiki/%EB%B0%98%EB%B3%B5%EC%9E%90)란 ?
객체 지향적 프로그래밍에서 배열이나 그와 유사한 자료구조 내부의 요소를 순회(traversing)하는 객체
- chain()
: 리스트( lists/tuples/iterables ) 를 연결하는 것
```python
letters = ['a', 'b', 'c', 'd', 'e', 'f']
booleans = [1, 0, 1, 0, 0, 1]
decimals = [0.1, 0.7, 0.4, 0.4, 0.5]
print(list(chain(letters, booleans, decimals)))
```
결과 : ['a', 'b', 'c', 'd', 'e', 'f', 1, 0, 1, 0, 0, 1, 0.1, 0.7, 0.4, 0.4, 0.5]
list()를 하지 않고 출력하게 되면 <itertools.chain object at 0x7f053ac09470> 이런식으로 출력됨.
- izip()
: 파이썬에서는 이미 요소를 튜플에 결합하는 zip () 함수가 표준 라이브러리에 있다.
izip()은 거의 같은 방식으로 작동하지만 약간의 성능 향상을 위해 iterable 객체를 반환한다. 

```python
print(list(izip([1, 2, 3], ['a', 'b', 'c'])))
```
결과 : [(1, 'a'), (2, 'b'), (3, 'c')]

이 외에도 count(), imap(), islice(), tee(), cycle(), repeat(), dropwhile(), takewhile(), ifilter(), groupby() --- 가 있습니다. 출처의 링크에 설명히 자세하게 나와있습니다. 

> 출처: https://hamait.tistory.com/803 [HAMA 블로그]

- 1등 코드에서의 사용(chain())
```python
layers = [['layer_1', 'layer_2', 'layer_3', 'layer_4'], [str(i) for i in np.arange(0, 226).tolist()]]
layers = list(chain(*layers))
```
![image](https://user-images.githubusercontent.com/44529556/106897854-d6870580-6736-11eb-82fc-ab5460089280.png)

layers앞에 *을 붙이지 않을 경우 unflattened result를 반환한다.
왜냐하면 chain은 각각의 list를 개별적으로 필요로 하는데 layers 통째로 하나의 항목으로 들어가기 때문이라고 한다.
*의 역할이 unpacking하여 함수에 전달하는 것 같다.

![image](https://user-images.githubusercontent.com/44529556/106898012-046c4a00-6737-11eb-8714-7823d2fe426d.png)
 .from_iterable을 붙이면 *와 같은 역할을 한다.

![image](https://user-images.githubusercontent.com/44529556/106898458-907e7180-6737-11eb-9c90-81b0a05c7561.png)

**Comment by mingxoxo** [comment]
<h2>DataLoader num_workers </h2>
num_workers은 학습 도중 CPU의 작업을 몇 개의 코어를 사용해서 진행할지에 대한 설정 파라미터

해당 환경에서 사용 가능한 코어의 개수를 확인해보고 학습 외의 작업에 영향을 주지 않을 정도의 숫자로 설정해주는 것이 좋다.

> 출처 : https://m.blog.naver.com/qbxlvnf11/221728476511

- 코드 사용 부분
```python
train_loader = DataLoader(train_dataset, batch_size=batch_size,  num_workers=4)
```

- colab에서 cpu 정보 출력
: 출력해보니 듀얼코어(processor 2개) 2.20GHz입니다.
따라서 제 코드에서 num_worker를 2로 바꿔서 실행하였습니다.

`!cat /proc/cpuinfo`
![image](https://user-images.githubusercontent.com/44529556/106908741-ec022c80-6742-11eb-9878-f3e362f42f53.png)

- colab에서 gpu 정보 출력

`!nvidia-smi`
![image](https://user-images.githubusercontent.com/44529556/106908848-063c0a80-6743-11eb-9faf-76bab05480a1.png)

**Comment by inh2613** [comment]
<h2>Skip connection </h2>

**[[출처]](https://lv99.tistory.com/25) Resnet ,Residual Learning**
-  Resnet모델에서 Kaming he 가 고안한 잔차(Residual)학습방법

- 레이어간의 연결이 순서대로 연속적인 것만 있는 것이 아니라, 중간을 뛰어넘어 전달하는(더하는) shortcut(=skip)이 추가됨. 
- 간단한 덧셈 연산으로 그래디언트를 직접적으로 잘 전달하는 역할을 하여 Gradient vanishing문제를 해결 

**Gradient vanishing**
- layer가 깊어질수록 미분을 점점 많이 하기 때문에 backpropagation을 해도 앞의 layer일수록 미분값이 작아져 그만큼 output에 영향을 끼치는 weight 정도가 작아지는 것

![다운로드](https://user-images.githubusercontent.com/62206617/106928944-1c53c600-6757-11eb-915b-717bcb52c7c0.png)



**코드사용부분 : 1등코드공유 class skipconnectionModule 선언부 안에 skipblock 부분**
```python
class skipConnectionModel(nn.Module):
    def __init__(self):
        super(skipConnectionModel, self).__init__()
        
        self.ln = nn.LayerNorm(10000)
        self.ln1 = nn.LayerNorm(7000)
        self.ln2 = nn.LayerNorm(4000)
        self.ln3 = nn.LayerNorm(2000)
        
        self.upblock1 = nn.Sequential(nn.Linear(226, 2000),nn.GELU(),nn.BatchNorm1d(2000))
        self.upblock2 = nn.Sequential(nn.Linear(2000,4000),nn.GELU(),nn.BatchNorm1d(4000))
        self.upblock3 = nn.Sequential(nn.Linear(4000,7000), nn.GELU(),nn.BatchNorm1d(7000))
        self.upblock4 = nn.Sequential(nn.Linear(7000,10000),nn.GELU(),nn.BatchNorm1d(10000))

        self.downblock1 = nn.Sequential(nn.Linear(10000, 7000),nn.GELU(),nn.BatchNorm1d(7000))
        self.downblock2 = nn.Sequential(nn.Linear(7000, 4000),nn.GELU(),nn.BatchNorm1d(4000))
        self.downblock3 = nn.Sequential(nn.Linear(4000, 2000),nn.GELU(),nn.BatchNorm1d(2000))
        self.downblock4 = nn.Sequential(nn.Linear(2000, 300),nn.GELU(),nn.BatchNorm1d(300))
        
        self.fclayer = nn.Sequential(nn.Linear(300,4))
        #fc layer 초기화 (초기화 방법이 중요하다 배웠었다.)
        #torch.nn.init.kaiming_uniform(self.fclayer.weight)
        self.dropout = nn.Dropout(0.1)
        
    def forward(self, x):
        upblock1_out = self.upblock1(x)
        upblock2_out = self.upblock2(upblock1_out)
        upblock3_out = self.upblock3(upblock2_out)
        upblock4_out = self.upblock4(upblock3_out)
        
        downblock1_out = self.downblock1(self.ln(upblock4_out))
        skipblock1 = downblock1_out + upblock3_out
        downblock2_out = self.downblock2(self.ln1(skipblock1))
        skipblock2 = downblock2_out + upblock2_out
        downblock3_out = self.downblock3(self.ln2(skipblock2))
        skipblock3 = downblock3_out + upblock1_out
        downblock4_out = self.downblock4(self.ln3(skipblock3))
        
        output = self.fclayer(downblock4_out)
        
        return output
```
1등 코드 공유에서의 모델의 경우,  downblock1_out의 결과에다가 input이었던 upblock3_out을 더함으로써 Skip connection을 하고 있습니다.

보다 자세한 설명: https://ganghee-lee.tistory.com/41 ,  https://leechamin.tistory.com/184

**Comment by mingxoxo** [comment]
<h2>Batch Normalization</h2>
각 feature의 평균과 분산을 구해서 batch에 있는 각 feature를 정규화 한다.
batch 전반에 걸쳐 처리되기 때문에 batch size와 관련이 깊다.

<h2>Layer Normalization</h2>
각 input의 feature들에 대한 평균과 분산을 구해서 batch에 있는 각 input를 정규화 한다.
각 input에 대해서만 처리되므로 batch size와 전혀 상관 없다.

![image](https://user-images.githubusercontent.com/44529556/106956170-91d08e00-6779-11eb-8a95-53a88699e5c2.png)
(그림에서 세로가 data기준 가로로 생각, 같은 row, 다른 feature)

![image](https://user-images.githubusercontent.com/44529556/106956501-03a8d780-677a-11eb-9ffa-4afe65e1d5d4.png)

> [출처](https://yonghyuc.wordpress.com/2020/03/04/batch-norm-vs-layer-norm/)

**Comment by inh2613** [comment]
<h2>learning rate Scheduling </h2>

Deap Learning  을 할 때, 학습 시 learning rate를 한 Step이 진행할 때마다 가변적으로 바꿀 수 있도록 하여(=학습률을 상황에 맞게 조절) Loss를 효과적으로 줄이는 방법

[참고 자료]  https://ai4nlp.tistory.com/8

Pytorch에서는 다양한 learning rate scheduler를 제공하고 있습니다.

[참고] https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate

**Warm up 스케줄링**

- 학습 초반에 **아주 작은** Learning rate로 시작하는 기법
- 초기학습이 안정적으로 수행됨


[참고] https://norman3.github.io/papers/docs/bag_of_tricks_for_image_classification.html

1등코드에서는 함수를 직접 짜서 learning rate 스케줄러를 만들었는데  기존에 있는 라이브러리를 사용한다면 

```python
import torch
import pytorch_warmup as warmup

optimizer = torch.optim.AdamW(params, lr=0.001, betas=(0.9, 0.999), weight_decay=0.01)
num_steps = len(dataloader) * num_epochs
lr_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=num_steps)
warmup_scheduler = warmup.UntunedLinearWarmup(optimizer)
for epoch in range(1,num_epochs+1):
    for batch in dataloader:
        optimizer.zero_grad()
        loss = ...
        loss.backward()
        optimizer.step()
        lr_scheduler.step()
        warmup_scheduler.dampen()
```
이렇게도 사용할 수 있을 것 같습니다!
[출처] https://pypi.org/project/pytorch-warmup/

찾아보니까 1등코드는 이곳에서 Learning rate 스케줄러랑 AdamW함수를 따온 거 같습니다.

https://huggingface.co/transformers/_modules/transformers/optimization.html#get_cosine_with_hard_restarts_schedule_with_warmup
